File: exercises/python/starting/checkout.py
from receipt import Receipt


class Checkout:
    def __init__(self):
        self.total = 0
        self.number_of_a = 0
        self.number_of_b = 0
        self.receipt = Receipt()

    def scan(self, sku):
        if sku == 'A':
            self.total += 50
            self.receipt.scanned_a()
        elif sku == 'B':
            self.total += 30
            self.receipt.scanned_b()
        elif sku == "C":
            self.total += 20
            self.receipt.scanned_c()
        elif sku == "D":
            self.total += 15
            self.receipt.scanned_d()

        if sku == 'A':
            self.number_of_a += 1
            if self.number_of_a % 3 == 0:
                self.total -= 20
        elif sku == "B":
            self.number_of_b += 1
            if self.number_of_b % 2 == 0:
                self.total -= 15

File: exercises/java/src/main/java/checkout/Checkout.java
package checkout;

import java.util.HashMap;
import java.util.Map;

class Checkout {

    private ProductLookup productLookup;
    private SpecialOffersLookup specialOffers;
    private Map<String, Integer> scannedItems;

    private int total;

    private Receipt receipt = new Receipt();


    public Checkout() {
        productLookup = new ProductLookup();
        specialOffers = new SpecialOffersLookup();
        scannedItems = new HashMap<>();
    }

    public Map<String, Integer> getScannedItems() {
        return scannedItems;
    }

    void scan(String sku) {

        // How much does this product cost?
        Product scannedProduct = productLookup.findProduct(sku);

        // Increase the quantity of products purchased
        if(scannedItems.containsKey(sku)) {
            int existingProductQuantity = scannedItems.get(sku);
            existingProductQuantity++;

            scannedItems.replace(sku, existingProductQuantity);
        } else {

            scannedItems.put(sku, 1);
        }

        // How much discount should be deducted from the total?
        Discount qualifyingDiscount = applyDiscounts(scannedProduct);

        adjustTotal(scannedProduct.getPrice(), qualifyingDiscount);

        receipt.addProduct(scannedProduct, qualifyingDiscount, total);

    }

    private Discount applyDiscounts(Product scannedProduct) {

        Discount qualifyingDiscount = null;

        for (Discount discount : specialOffers.getSpecialOffersLookup()) {
            if (scannedProduct.getSku().equals(discount.getSku())) {
                if(scannedItems.get(scannedProduct.getSku()) % discount.getQualifyingQuantity() == 0) {
                    qualifyingDiscount = discount;
                }
            }
        }

        return qualifyingDiscount;

    }

    void adjustTotal(int productAmount, Discount qualifyingDiscount) {

        if(qualifyingDiscount != null) {
            total += (productAmount - qualifyingDiscount.getAmountToDeduct());
        } else {
            total += productAmount;
        }

    }

    int total() {
        return total;
    }

    public String receipt() {
        return receipt.printReceipt(total);
    }
}
